python/pyspark/sql/connect/client/reattach.py

```diff
@@ -0,0 +1,228 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from pyspark.sql.connect.utils import check_dependencies
+
+check_dependencies(__name__)
+
+import uuid
+from collections.abc import Generator
+from typing import Optional, Dict, Any, Iterator
+import threading
+
+import pyspark.sql.connect.proto as pb2
+import pyspark.sql.connect.proto.base_pb2_grpc as grpc_lib
+
+
+class ExecutePlanResponseReattachableIterator(Generator):
+    """
+    Retryable iterator of ExecutePlanResponses to an ExecutePlan call.
+
+    It can handle situations when:
+      - the ExecutePlanResponse stream was broken by retryable network error (governed by
+        retryPolicy)
+      - the ExecutePlanResponse was gracefully ended by the server without a ResultComplete
+        message; this tells the client that there is more, and it should reattach to continue.
+
+    Initial iterator is the result of an ExecutePlan on the request, but it can be reattached with
+    ReattachExecute request. ReattachExecute request is provided the responseId of last returned
+    ExecutePlanResponse on the iterator to return a new iterator from server that continues after
+    that.
+
+    Since in reattachable execute the server does buffer some responses in case the client needs to
+    backtrack
+    """
+
+    def __init__(
+        self,
+        request: pb2.ExecutePlanRequest,
+        stub: grpc_lib.SparkConnectServiceStub,
+        retry_policy: Dict[str, Any],
+    ):
+        self._request = request
+        self._retry_policy = retry_policy
+        if request.operation_id:
+            self._operation_id = request.operation_id
+        else:
+            # Add operation id, if not present.
+            # with operationId set by the client, the client can use it to try to reattach on error
+            # even before getting the first response. If the operation in fact didn't even reach the
+            # server, that will end with INVALID_HANDLE.OPERATION_NOT_FOUND error.
+            self._operation_id = str(uuid.uuid4())
+
+        self._stub = stub
+        request.request_options.append(
+            pb2.ExecutePlanRequest.RequestOption(
+                reattach_options=pb2.ReattachOptions(reattachable=True)
+            )
+        )
+        self._initial_request = request
+
+        # ResponseId of the last response returned by next()
+        self._last_returned_response_id: Optional[str] = None
+
+        # True after ResponseComplete message was seen in the stream.
+        # Server will always send this message at the end of the stream, if the underlying iterator
+        # finishes without producing one, another iterator needs to be reattached.
+        self._result_complete = False
+
+        # Initial iterator comes from ExecutePlan request.
+        self._iterator: Iterator[pb2.ExecutePlanResponse] = self._stub.ExecutePlan(
```

Comment:
So in the case of the target resource throwing a retriable error, this will already raise an exception and break the retry logic. It looks like the current retry handling can only deal with subsequent errors in the handling of the response stream but not the initial error itself.

Reply:
@juliuszsompolski is this the same for the current scala client?

Reply:
At least in scala, calling the `executePlan` never throws error, even if the service is absolutely not available and one could expect immediate error; I tested that with completely killing the server, then calling executePlan, and never getting an actual error before trying to call next()... The error is only thrown on the first `next()` or `hasNext()` call of the resulting iterator. So in scala not catching and retrying here, but at the first iterator invocation is fine. So in https://github.com/apache/spark/pull/42228/files#diff-2efc03ab3edd5e5f78652b1454b31fd4917113de09a64fbdc7352f150c685628R89 and also in https://github.com/apache/spark/pull/41829/files#diff-79d952ace19b8b9d4b04e010653aba35e4aeaca957946cc68414b7a766deea99R68 doesn't wrap the first call
we do not retry.

But it's a good callout that it needs checking if python GRPC behaves the same, or maybe just defensively wrap it in retry anyway (and do the same in scala just in case it's a GRPC internal quirk that we shouldn't depend on), or at least it deserves a comment.

Reply:
I raised https://github.com/apache/spark/pull/42274 to be defensive about it in scala.

Reply:
... but what I write here implies that if the error was indeed during the initial ExecutePlan that didn't reach the server, and we catch it on the first hasNext() later, then Reattach won't work, because the query won't be on the server, so it will fail with INVALID_HANDLE.OPERATION_NOT_FOUND. So this indeed means that the case it won't work if retry of the initial ExecutePlan is needed, and wrapping the retry here won't help for that.

Reply:
retrying the failure of initial operation starting is tricky, because:
- if the initial operation reached the server, retry of ExecutePlan would fail with INVALID_HANDLE.OPERATION_ALREADY_EXISTS; in that case ReattachPlan should be used. And this is what happens here.
- but if the initial operation didn't reach the server, then the ReattachPlan will fail with INVALID_HANDLE.OPERATION_NOT_FOUND. This could be improved by making it switch to retrying the ExecutePlan then
But I think this is a border case that can be a followup or decided it's not worth it.

I believe in the state as is, we will get an INVALID_HANDLE.OPERATION_NOT_FOUND error in case the original ExecutePlan never reached the server. In that case, we could improve it by trying to call ExecutePlan again. But that would have to be caught not here, but during the next() calls.

Reply:
So now I have 3 PRs:
* https://github.com/apache/spark/pull/42281 adds some comments in scala to explain that GRPC throws errors deferred to the iterator functions, and the limitation of retry of the original execute didn't reach the server
* https://github.com/apache/spark/pull/42282 fixes the retry in case original execute didn't reach the server
* https://github.com/apache/spark/pull/42274 does that plus does overkill retries in case the assumption that calling executePlan and reattachExecute won't throw is wrong.

I would suggest merging the first, and discussing this corner case later.